File: mobile/ios/Sources/UI/RelativeTime.swift
import Foundation

// Pure formatting, deliberately free of any View/Date-now dependency so it
// is trivial to test at exact minute boundaries. Two shapes are needed:
// the compact form the connection line uses next to the host ("2m", "3h"),
// and the full sentence DESIGN.md's copy rules specify for a stale widget
// ("Last synced 3h ago").
public enum RelativeTime {
    // Connection line: "<dot> host · <compact>".
    public static func compact(minutes: Int) -> String {
        let minutes = max(0, minutes)
        if minutes == 0 { return "now" }
        if minutes < 60 { return "\(minutes)m" }
        let hours = minutes / 60
        if hours < 24 { return "\(hours)h" }
        return "\(hours / 24)d"
    }

    // DESIGN.md: "Last synced 3h ago" — state the fact, do not say "offline"
    // when that isn't known.
    public static func lastSynced(minutes: Int) -> String {
        let minutes = max(0, minutes)
        if minutes == 0 { return "Last synced just now" }
        return "Last synced \(compact(minutes: minutes)) ago"
    }

    // TodayView / SettingsView's fuller "Updated 2m ago" status line.
    public static func updated(minutes: Int) -> String {
        let minutes = max(0, minutes)
        return minutes == 0 ? "Updated just now" : "Updated \(compact(minutes: minutes)) ago"
    }
}

File: mobile/android/app/src/main/kotlin/sh/openagi/mobile/util/RelativeTime.kt
package sh.openagi.mobile.util

// Pure, deliberately: the connection line and the widget both render an age
// in minutes into the same short phrase, and this is the one place that
// phrasing is decided. No Android/Instant dependency so it is trivial to
// pin with a table of (minutes -> expected string) tests.
object RelativeTime {
    // "just now", "2m", "1h", "3h", "2d" — the compact form next to a host.
    fun short(ageMinutes: Int): String = when {
        ageMinutes <= 0 -> "just now"
        ageMinutes < 60 -> "${ageMinutes}m"
        ageMinutes < 60 * 24 -> "${ageMinutes / 60}h"
        else -> "${ageMinutes / (60 * 24)}d"
    }

    // "Updated just now" / "Updated 2m ago" — TodayScreen's status line.
    fun updated(ageMinutes: Int): String =
        if (ageMinutes <= 0) "Updated just now" else "Updated ${short(ageMinutes)} ago"

    // "Last synced 3h ago" — DESIGN.md's exact stale-widget copy, reused
    // wherever a screen or the widget must say a snapshot has gone stale
    // without claiming to know the daemon is offline.
    fun lastSynced(ageMinutes: Int): String = "Last synced ${short(ageMinutes)} ago"
}
